### DISCUSSION TITLE
How would you store global settings?

### DISCUSSION THREAD

**Discussion by ghost** [isar/isar#1066] [Help]
Can Isar store objects like in a dictionary (key, value)? Or only collections (arrays)? How would you store user settings like dark mode or last selected id from a collection, etc. Currently I have a settings collection holding on index 0 (id 1).

**Comment by MrCsabaToth** [comment]
I'll migrate my app to isar and besides my SQLite database I'll migrate the Android preferences into Isar as well. This way a full JSON export will contain the settings as well so the whole shebang can be saved and restored by the user.

As far as I understand Isar is relational. Android preferences has a few simple types, so what I'm planning now is having an Isar collection of something like this:

```
@Collection(inheritance: false)
class SettingsEntry {
  Id id;
  @Index()
  String name;
  String? stringValue;
  int? intValue;
  double? doubleValue;
  bool? boolValue;
}
```

These values will be either nullable or initialized with a well selected nonsense value. Optionally I may add a `kind` field which explicitly tells what type is the settings entry instead of figuring it out by which value field is filled.

Then there will be a manager class which will provide a simple settings API to get and set these values. Since the collection is indexed by name the lookup should be almost as efficient as a dictionary. It would be great if this would be some kind of a plugin if isar.

Another hurdle will be to migrate the UX portion of the settings which is right now based on https://pub.dev/packages/pref ... unless I can formulate an API which is equivalent of the preferences and then I can somehow straight-up keep using `pref`.

I'm planning to migrate my debug logging facility as well, which is right now based on https://pub.dev/packages/flutter_logs, but there's some issue with it and isar could be an improvement over it.

Let me know what others think and if I'm wrong.

**Comment by ghost** [comment]
Interesting idea. All solutions have pros and cons. The advantages here is you can handle/update each setting separately. I wonder if Isar can handle dynamic values. So you only have `name` and `value`. It would clean up the object structure. But `any` type ala JavaScript is maybe not the best solution. Maybe better with generic type? Idk. Anyway... My current solution is to hold one object on id 1. So I just have to load it once at app startup and can access all values in a sync way. But I have to save/update the whole object on each setting change. Pros/cons. I'm new in Flutter/Dart/Isar btw. It looks like:

```dart
@collection
class Settings {
  Id id = Isar.autoIncrement;
  bool darkMode = false;
  int anotherValue = 0;
  // etc
}
```

```dart
// Load once on app init.
final settings = await isar.settings.get(1);
// Save
// Update (id 1)
```

I craft my own settings page. Libs like pref can be helpful and for some devs or project the right choice. But I'm fine with my own implementation with full control in function and style. I think I'll create some option type widgets and a settings accessor.

Btw. why `inheritance: false`? I know the [docs](https://isar.dev/schema.html#anatomy-of-a-collection), but I don't fully understand at this moment and context. What's the parent class or mixin in this case?

> Control whether fields of parent classes and mixins will be stored in Isar. Enabled by default.

**Reply by MrCsabaToth** [reply]
Interesting! This is also a viable route, and the API you build around the entity can use `dynamic`. IN my case however I have a whole bunch of settings (palette color settings for light/dark and background/foreground for 5, 6, and 7 zone palettes which are handled in a double loop, I would not spell those out one-by-one.

I put the `inheritance: false` because I knew I won't use any inheritance (from persistence perspective) and I thought it may help the code generation to know that.

**Reply by MrCsabaToth** [reply]
One more note: your proposal depends on how Isar handles schema changes. Each time there's a new setting a schema change need to be made. I haven't experienced that how that goes with Isar yet. I'm migrating from Floor, and I'm writing explicit migrations (in SQL lite) there. BTW I also discovered that https://pub.dev/packages/pref has a [toMap](https://github.com/DavBfr/flutter_pref/blob/master/pref/lib/src/service/base.dart#L113) and [fromMap](https://github.com/DavBfr/flutter_pref/blob/master/pref/lib/src/service/base.dart#L122) on its API and maybe I just stay with `SharedPreferences`. Because I don't want to rewrite my whole extensive settings UI, otherwise I'd need to write some kind of an API wrapper which would use Isar under the hood, and I'm not sure if I want to spend time on that.
